crypto/hpke/hpke.c

```diff
@@ -0,0 +1,2579 @@
+        return 0;
+    return 1;
+err:
+    return erv;
+}
+
+/*
+ * @brief string matching for suites
+ */
+#if defined(_WIN32)
+# define HPKE_MSMATCH(inp, known) \
+    (strlen(inp) == strlen(known) && !_stricmp(inp, known))
+#else
+# define HPKE_MSMATCH(inp, known) \
+    (strlen(inp) == strlen(known) && !strcasecmp(inp, known))
+#endif
+
+/*
+ * @brief map a string to a HPKE suite
+ *
+ * @param str is the string value
+ * @param suite is the resulting suite
+ * @return 1 for success, otherwise failure
+ */
+static int hpke_str2suite(const char *suitestr, OSSL_HPKE_SUITE *suite)
+{
+    uint16_t kem = 0, kdf = 0, aead = 0;
+    char *st = NULL;
+    char *instrcp = NULL;
+    size_t inplen = 0;
+    int labels = 0;
+
+    if (suitestr == NULL || suite == NULL)
+        return 0;
+    /* See if it contains a mix of our strings and numbers  */
+    inplen = OPENSSL_strnlen(suitestr, OSSL_HPKE_MAX_SUITESTR);
+    if (inplen >= OSSL_HPKE_MAX_SUITESTR)
+        return 0;
+    instrcp = OPENSSL_strndup(suitestr, inplen);
+    st = strtok(instrcp, ",");
+    if (st == NULL) {
+        OPENSSL_free(instrcp);
+        return 0;
+    }
+    while (st != NULL && ++labels <= 3) {
+        /* check if string is known or number and if so handle appropriately */
+        if (kem == 0) {
+            if (HPKE_MSMATCH(st, OSSL_HPKE_KEMSTR_P256)) {
+                kem = OSSL_HPKE_KEM_ID_P256;
+            }
+            if (HPKE_MSMATCH(st, OSSL_HPKE_KEMSTR_P384)) {
+                kem = OSSL_HPKE_KEM_ID_P384;
+            }
+            if (HPKE_MSMATCH(st, OSSL_HPKE_KEMSTR_P521)) {
+                kem = OSSL_HPKE_KEM_ID_P521;
+            }
+            if (HPKE_MSMATCH(st, OSSL_HPKE_KEMSTR_X25519)) {
+                kem = OSSL_HPKE_KEM_ID_X25519;
+            }
+            if (HPKE_MSMATCH(st, OSSL_HPKE_KEMSTR_X448)) {
+                kem = OSSL_HPKE_KEM_ID_X448;
+            }
+            if (HPKE_MSMATCH(st, "0x10")) { kem = OSSL_HPKE_KEM_ID_P256; }
+            if (HPKE_MSMATCH(st, "16")) { kem = OSSL_HPKE_KEM_ID_P256; }
+            if (HPKE_MSMATCH(st, "0x11")) { kem = OSSL_HPKE_KEM_ID_P384; }
+            if (HPKE_MSMATCH(st, "17")) { kem = OSSL_HPKE_KEM_ID_P384; }
+            if (HPKE_MSMATCH(st, "0x12")) { kem = OSSL_HPKE_KEM_ID_P521; }
+            if (HPKE_MSMATCH(st, "18")) { kem = OSSL_HPKE_KEM_ID_P521; }
+            if (HPKE_MSMATCH(st, "0x20")) { kem = OSSL_HPKE_KEM_ID_X25519; }
+            if (HPKE_MSMATCH(st, "32")) { kem = OSSL_HPKE_KEM_ID_X25519; }
+            if (HPKE_MSMATCH(st, "0x21")) { kem = OSSL_HPKE_KEM_ID_X448; }
+            if (HPKE_MSMATCH(st, "33")) { kem = OSSL_HPKE_KEM_ID_X448; }
+        } else if (kem != 0 && kdf == 0) {
+            if (HPKE_MSMATCH(st, OSSL_HPKE_KDFSTR_256)) { kdf = 1; }
+            if (HPKE_MSMATCH(st, OSSL_HPKE_KDFSTR_384)) { kdf = 2; }
+            if (HPKE_MSMATCH(st, OSSL_HPKE_KDFSTR_512)) { kdf = 3; }
+            if (HPKE_MSMATCH(st, "0x01")) { kdf = 1; }
```

Comment:
Maybe dont use magic numbers here for kdf= and aead =

Reply:
fair cop - fixed